lib/puppet/type/cisco_ace.rb

```diff
@@ -0,0 +1,189 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'ipaddr'
+begin
+  require 'puppet_x/cisco/cmnutils'
+rescue LoadError # seen on master, not on agent
+  # See longstanding Puppet issues #4248, #7316, #14073, #14149, etc. Ugh.
+  require File.expand_path(File.join(File.dirname(__FILE__), '..', '..',
+                                     'puppet_x', 'cisco', 'cmnutils.rb'))
+end
+
+Puppet::Type.newtype(:cisco_ace) do
+  @doc = "Manages ACE configuration.
+
+  ~~~puppet
+  cisco_ace { '<afi> <acl_name> <seqno>':
+    ..attributes..
+  }
+  ~~~
+
+  `<afi> <acl_name> <seqno>` is the title of the ace resource.
+
+  Example:
+
+  ~~~puppet
+    cisco_ace { 'ipv4 acl-foo-1 10':
+      action                                => permit,
+      proto                                 => tcp,
+      src_addr                              => 1.2.3.4 2.3.4.5,
+      src_port                              => eq 40,
+      dst_addr                              => 8.9.0.4/32,
+      dst_port                              => range 32 56,
+    }
+  ~~~
+  "
+
+  ###################
+  # Resource Naming #
+  ###################
+  # Parse out the title to fill in the attributes in these patterns. These
+  # attributes can be overwritten later.
+
+  def self.title_patterns
+    identity = ->(x) { x }
+    [
+      [
+        /^(ipv4|ipv6)\s+(\S+)\s+(\d+)$/,
+        [
+          [:afi, identity],
+          [:acl_name, identity],
+          [:seqno, identity],
+        ],
+      ]
+    ]
+  end
+
+  ##############
+  # Parameters #
+  ##############
+
+  ensurable
+
+  # Overwrites the acl_name method which by default returns only
+  # self[:name].
+  def name
+    "#{self[:afi]} #{self[:acl_name]} #{self[:seqno]}"
+  end
+
+  # Only needed to satisfy name parameter.
+  newparam(:name) do
+  end
+
+  newparam(:afi, namevar: true) do
+    desc 'The Address-Family Indentifier (ipv4|ipv6).'
+    newvalues(:ipv4, :ipv6)
+  end
+
+  newparam(:acl_name, namevar: true) do
+    desc 'acl_name '
+  end
+
+  newparam(:seqno, namevar: true) do
+    desc 'seqno of ace'
+  end
+
+  ##############
+  # Properties #
+  ##############
+
+  newproperty(:action) do
+    desc 'Ace Action Identifier (permit|deny)'
+
+    validate do |action|
+      fail 'action should be permit or deny ' unless
+        /permit|deny/.match(action.downcase)
+    end
+
+    munge(&:downcase)
+  end
+
+  newproperty(:proto) do
+    desc 'protocol of ace'
+
+    validate do |proto|
+      fail 'proto must be type String or Integer' unless
+        /\S+|\d+/.match(proto)
+    end
+
+    munge do |protocol|
+      protocol
+    end
+  end
+
+  newproperty(:src_addr) do
+    desc 'src address of the ace'
```

Comment:
Please review *all* of the `desc` and `fail` messages. The `desc` messages should state what the property is and what the expected inputs should be, ie:

```
    desc "Source address to match against. Valid values are an IP address/prefix_len (10.0.0.0/8), '\
        "an address group (addrgroup foo), or the keyword 'any'"
```

The `fail` messages should be similar or exactly the same if it's appropriate.